### PR TITLE
Add LHEPtFilter plugin needed for MC generation 

### PR DESCRIPTION
**Pull request by ajgilbert**
#### PR description:

Adds a new plugin in `GeneratorInterface/GenFilters` that applies a user specified pT (or pT range) cut  to the p4 sum of a subset of LHE-level particles. This is needed specifically for the production of Wgamma samples binned in the photon pT, but may be useful more generally in the future.

#### PR validation:

All tests in `runTheMatrix.py -l limited -i all --ibeos` passed.

#### if this PR is a backport please specify the original PR:

Not a backport.

### CODE REVIEW

**Comment by fabiocos** [comment]
@ajgilbert do you really need a separate header?

**Comment by ajgilbert** [comment]
@fabiocos Indeed not - have removed it.

**Comment by fabiocos** [comment]
@ajgilbert I assume you have privately tested and this has been scrutinized by GEN conveners, do you have some validation documentation to link?

**Comment by fabiocos** [comment]
please test

File: GeneratorInterface/GenFilters/plugins/LHEPtFilter.cc [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+#include "FWCore/Framework/interface/MakerMacros.h"
+
+#include "FWCore/ParameterSet/interface/ParameterSet.h"
+#include "SimDataFormats/GeneratorProducts/interface/LHEEventProduct.h"
+
+//
+// class declaration
+//
+
+class LHEPtFilter : public edm::EDFilter {
```
Comment (makortel) [inline GeneratorInterface/GenFilters/plugins/LHEPtFilter.cc]: Could this be made a non-legacy module, preferably global or stream?

File: GeneratorInterface/GenFilters/plugins/LHEPtFilter.cc [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+
+private:
+  bool filter(edm::Event&, const edm::EventSetup&) override;
+  void endJob() override;
+
+  // ----------member data ---------------------------
+
+  edm::EDGetTokenT<LHEEventProduct> src_;
+  std::vector<lhef::HEPEUP::FiveVector> lheParticles;
+  std::vector<ROOT::Math::PxPyPzEVector> cands;
```
Comment (makortel) [inline GeneratorInterface/GenFilters/plugins/LHEPtFilter.cc]: By quick look it looks like `lheParticles` and `cands` are not needed as member variables.

File: GeneratorInterface/GenFilters/plugins/LHEPtFilter.cc (hunk 1 above)
Comment (ajgilbert) [reply GeneratorInterface/GenFilters/plugins/LHEPtFilter.cc]: Yes, has been converted to a global module in the latest commits.

File: GeneratorInterface/GenFilters/plugins/LHEPtFilter.cc (hunk 2 above)
Comment (ajgilbert) [reply GeneratorInterface/GenFilters/plugins/LHEPtFilter.cc]: Right, have switched these to local variables now.

**Comment by fabiocos** [comment]
@ajgilbert @makortel does this need to be a global module? For what I can see it might work as a stream module, there is no global action implied, only event-by-event.

**Comment by fabiocos** [comment]
please test

**Comment by agrohsje** [comment]
+1

**Comment by makortel** [comment]
@fabiocos 

> does this need to be a global module? For what I can see it might work as a stream module, there is no global action implied, only event-by-event.

"Global" is stricter category for modules than "stream" (must be immutable or thread safe). Framework creates one instance of a global module (wrt. an instance per stream for stream modules) leading to smaller memory consumption for immutable modules. Therefore it is preferable to use global modules where feasible.

**Comment by fabiocos** [comment]
@makortel thanks for the comment, global is stricter indeed, my comment was about the fact that was not strictly needed by the kind of operation performed in the code as it is. I agree that this implementation is more robust against a possible update that would require global operations.

**Comment by fabiocos** [comment]
+1

**Pull request merged**